File: platform/services/catalog/internal/pricing/pricing_test.go
package pricing

import "testing"

// Hungarian menu prices are gross: the customer sees 1270 HUF and 27% VAT is
// already inside it. HUF has no minor unit, so these are whole forint.
func TestPriceLine_TaxIncluded_HUF(t *testing.T) {
	got, err := PriceLine(Line{
		Quantity: 1, UnitPriceMinor: 1270, Currency: "HUF",
		TaxBasisPoints: 2700, TaxIncluded: true,
	})
	if err != nil {
		t.Fatal(err)
	}
	// 1270 / 1.27 = 1000 exactly.
	if got.GrossMinor != 1270 || got.NetMinor != 1000 || got.TaxMinor != 270 {
		t.Fatalf("got %+v, want gross=1270 net=1000 tax=270", got)
	}
}

func TestPriceLine_TaxExcluded(t *testing.T) {
	got, err := PriceLine(Line{
		Quantity: 2, UnitPriceMinor: 500, Currency: "BDT",
		TaxBasisPoints: 1500, TaxIncluded: false,
	})
	if err != nil {
		t.Fatal(err)
	}
	// net 1000, tax 15% = 150, gross 1150
	if got.NetMinor != 1000 || got.TaxMinor != 150 || got.GrossMinor != 1150 {
		t.Fatalf("got %+v, want net=1000 tax=150 gross=1150", got)
	}
}

// The rounding rule is the point of this test: truncation would return 236.
func TestPriceLine_RoundsHalfUp(t *testing.T) {
	got, err := PriceLine(Line{
		Quantity: 1, UnitPriceMinor: 999, Currency: "HUF",
		TaxBasisPoints: 2700, TaxIncluded: true,
	})
	if err != nil {
		t.Fatal(err)
	}
	// 999 * 10000 / 12700 = 786.61… -> 787 net, so tax = 212.
	if got.NetMinor != 787 || got.TaxMinor != 212 {
		t.Fatalf("got net=%d tax=%d, want net=787 tax=212", got.NetMinor, got.TaxMinor)
	}
	if got.NetMinor+got.TaxMinor != got.GrossMinor {
		t.Fatalf("net+tax != gross: %+v", got)
	}
}

// Whatever the rounding, a line must always reconcile internally.
func TestPriceLine_NetPlusTaxAlwaysEqualsGross(t *testing.T) {
	for price := int64(1); price <= 3000; price++ {
		for _, rate := range []int32{0, 500, 1800, 2700} {
			a, err := PriceLine(Line{
				Quantity: 1, UnitPriceMinor: price, Currency: "HUF",
				TaxBasisPoints: rate, TaxIncluded: true,
			})
			if err != nil {
				t.Fatal(err)
			}
			if a.NetMinor+a.TaxMinor != a.GrossMinor {
				t.Fatalf("price=%d rate=%d: net %d + tax %d != gross %d",
					price, rate, a.NetMinor, a.TaxMinor, a.GrossMinor)
			}
		}
	}
}

func TestPriceLine_Discount(t *testing.T) {
	got, err := PriceLine(Line{
		Quantity: 2, UnitPriceMinor: 1000, Currency: "HUF",
		TaxBasisPoints: 2700, TaxIncluded: true, DiscountMinor: 500,
	})
	if err != nil {
		t.Fatal(err)
	}
	if got.GrossMinor != 1500 {
		t.Fatalf("gross = %d, want 1500 (2000 less 500 discount)", got.GrossMinor)
	}
}

func TestPriceLine_Rejects(t *testing.T) {
	cases := map[string]Line{
		"zero quantity":    {Quantity: 0, UnitPriceMinor: 100, Currency: "HUF"},
		"unknown currency": {Quantity: 1, UnitPriceMinor: 100, Currency: "XYZ"},
		"discount > line":  {Quantity: 1, UnitPriceMinor: 100, Currency: "HUF", DiscountMinor: 500},
		"absurd tax rate":  {Quantity: 1, UnitPriceMinor: 100, Currency: "HUF", TaxBasisPoints: 200_000},
	}
	for name, l := range cases {
		if _, err := PriceLine(l); err == nil {
			t.Errorf("%s: expected an error, got none", name)
		}
	}
}

// Summing rounded lines is what makes a receipt add up. Deriving tax from the
// gross total instead can differ by a unit or two — enough to fail an audit.
func TestTotal_SumsRoundedLines(t *testing.T) {
	var as []Amounts
	for i := 0; i < 3; i++ {
		a, err := PriceLine(Line{
			Quantity: 1, UnitPriceMinor: 999, Currency: "HUF",
			TaxBasisPoints: 2700, TaxIncluded: true,
		})
		if err != nil {
			t.Fatal(err)
		}
		as = append(as, a)
	}
	tot, err := Total(as)
	if err != nil {
		t.Fatal(err)
	}
	if tot.GrossMinor != 2997 || tot.NetMinor != 2361 || tot.TaxMinor != 636 {
		t.Fatalf("got %+v, want gross=2997 net=2361 tax=636", tot)
	}
	if tot.NetMinor+tot.TaxMinor != tot.GrossMinor {
		t.Fatalf("total does not reconcile: %+v", tot)
	}
}

func TestTotal_RejectsMixedCurrency(t *testing.T) {
	_, err := Total([]Amounts{{Currency: "HUF"}, {Currency: "BDT"}})
	if err == nil {
		t.Fatal("expected mixed-currency total to be rejected")
	}
}

File: platform/services/web/packages/money/src/pricing.test.ts
/**
 * Ported from platform/services/catalog/internal/pricing/pricing_test.go.
 *
 * These are the same cases with the same expected figures. They exist twice on
 * purpose: the till prices a cart with this code and the receipt is issued by
 * that code, and a customer holding the printout next to the screen is
 * entitled to see one number. If one suite is changed without the other, that
 * guarantee is gone and nothing else will notice.
 */
import { describe, expect, it } from 'vitest'
import { money, MoneyError } from './money'
import { priceLine, PricingError, taxBreakdown, totalOf, type Amounts } from './pricing'

const huf = (amount: number) => money(amount, 'HUF')
const bdt = (amount: number) => money(amount, 'BDT')

describe('priceLine', () => {
  // Hungarian menu prices are gross: the customer sees 1270 HUF and 27% VAT is
  // already inside it. HUF has no minor unit, so these are whole forint.
  it('derives net from a tax-inclusive HUF price', () => {
    const amounts = priceLine({
      quantity: 1,
      unitPrice: huf(1270),
      taxBasisPoints: 2700,
      taxIncluded: true,
    })
    // 1270 / 1.27 = 1000 exactly.
    expect(amounts.gross.minor).toBe(1270)
    expect(amounts.net.minor).toBe(1000)
    expect(amounts.tax.minor).toBe(270)
  })

  it('adds tax to a tax-exclusive price', () => {
    const amounts = priceLine({
      quantity: 2,
      unitPrice: bdt(500),
      taxBasisPoints: 1500,
      taxIncluded: false,
    })
    expect(amounts.net.minor).toBe(1000)
    expect(amounts.tax.minor).toBe(150)
    expect(amounts.gross.minor).toBe(1150)
  })

  // The rounding rule is the point of this test: truncation returns 786.
  it('rounds half away from zero', () => {
    const amounts = priceLine({
      quantity: 1,
      unitPrice: huf(999),
      taxBasisPoints: 2700,
      taxIncluded: true,
    })
    // 999 * 10000 / 12700 = 786.61... so net is 787 and tax is 212.
    expect(amounts.net.minor).toBe(787)
    expect(amounts.tax.minor).toBe(212)
    expect(amounts.net.minor + amounts.tax.minor).toBe(amounts.gross.minor)
  })

  // Whatever the rounding, a line must always reconcile internally.
  it('always reconciles net plus tax to gross', () => {
    for (let price = 1; price <= 3000; price++) {
      for (const rate of [0, 500, 1800, 2700]) {
        const amounts = priceLine({
          quantity: 1,
          unitPrice: huf(price),
          taxBasisPoints: rate,
          taxIncluded: true,
        })
        expect(
          amounts.net.minor + amounts.tax.minor,
          `price ${price} at ${rate}bp`,
        ).toBe(amounts.gross.minor)
      }
    }
  })

  it('applies a discount before tax is derived', () => {
    const amounts = priceLine({
      quantity: 2,
      unitPrice: huf(1000),
      taxBasisPoints: 2700,
      taxIncluded: true,
      discount: huf(500),
    })
    expect(amounts.gross.minor).toBe(1500)
  })

  it('rejects a zero quantity', () => {
    expect(() =>
      priceLine({ quantity: 0, unitPrice: huf(100), taxBasisPoints: 0, taxIncluded: true }),
    ).toThrow(PricingError)
  })

  it('rejects a fractional quantity', () => {
    expect(() =>
      priceLine({ quantity: 1.5, unitPrice: huf(100), taxBasisPoints: 0, taxIncluded: true }),
    ).toThrow(PricingError)
  })

  it('rejects an unknown currency', () => {
    // Construction refuses it first, which is the boundary that matters.
    expect(() => money(100, 'XYZ')).toThrow(MoneyError)
    // And priceLine refuses it too, for a value that reached it another way.
    const smuggled = { minor: 100, currency: 'XYZ' } as ReturnType<typeof huf>
    expect(() =>
      priceLine({ quantity: 1, unitPrice: smuggled, taxBasisPoints: 0, taxIncluded: true }),
    ).toThrow(PricingError)
  })

  it('rejects a discount larger than the line', () => {
    expect(() =>
      priceLine({
        quantity: 1,
        unitPrice: huf(100),
        taxBasisPoints: 0,
        taxIncluded: true,
        discount: huf(500),
      }),
    ).toThrow(PricingError)
  })

  it('rejects an absurd tax rate', () => {
    expect(() =>
      priceLine({
        quantity: 1,
        unitPrice: huf(100),
        taxBasisPoints: 200_000,
        taxIncluded: true,
      }),
    ).toThrow(PricingError)
  })

  it('rejects a discount in a different currency from the line', () => {
    expect(() =>
      priceLine({
        quantity: 1,
        unitPrice: huf(1000),
        taxBasisPoints: 2700,
        taxIncluded: true,
        discount: bdt(100),
      }),
    ).toThrow(PricingError)
  })
})

describe('totalOf', () => {
  // Summing rounded lines is what makes a receipt add up. Deriving tax from the
  // gross total instead can differ by a unit or two, which is enough to fail
  // an audit.
  it('sums rounded lines rather than re-deriving from the total', () => {
    const lines: Amounts[] = Array.from({ length: 3 }, () =>
      priceLine({
        quantity: 1,
        unitPrice: huf(999),
        taxBasisPoints: 2700,
        taxIncluded: true,
      }),
    )
    const total = totalOf(lines)
    expect(total.gross.minor).toBe(2997)
    expect(total.net.minor).toBe(2361)
    expect(total.tax.minor).toBe(636)
    expect(total.net.minor + total.tax.minor).toBe(total.gross.minor)
  })

  it('rejects a mixed-currency total', () => {
    const hufLine = priceLine({
      quantity: 1,
      unitPrice: huf(100),
      taxBasisPoints: 0,
      taxIncluded: true,
    })
    const bdtLine = priceLine({
      quantity: 1,
      unitPrice: bdt(100),
      taxBasisPoints: 0,
      taxIncluded: true,
    })
    expect(() => totalOf([hufLine, bdtLine])).toThrow(PricingError)
  })

  it('rejects an empty total', () => {
    expect(() => totalOf([])).toThrow(PricingError)
  })
})

describe('taxBreakdown', () => {
  it('groups by rate and sorts ascending', () => {
    const breakdown = taxBreakdown([
      { quantity: 1, unitPrice: huf(1270), taxBasisPoints: 2700, taxIncluded: true },
      { quantity: 1, unitPrice: huf(1050), taxBasisPoints: 500, taxIncluded: true },
      { quantity: 2, unitPrice: huf(1270), taxBasisPoints: 2700, taxIncluded: true },
    ])
    expect(breakdown.map((band) => band.basisPoints)).toEqual([500, 2700])
    // 1270 + 2540 gross at 27%.
    expect(breakdown[1]?.gross.minor).toBe(3810)
  })
})
